presto-main/src/main/java/io/prestosql/sql/planner/optimizations/RemoveRedundantOrderByOptimizer.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.optimizations;
+
+import io.airlift.log.Logger;
+import io.prestosql.Session;
+import io.prestosql.execution.warnings.WarningCollector;
+import io.prestosql.sql.planner.PlanNodeIdAllocator;
+import io.prestosql.sql.planner.SymbolAllocator;
+import io.prestosql.sql.planner.TypeProvider;
+import io.prestosql.sql.planner.plan.AggregationNode;
+import io.prestosql.sql.planner.plan.ChildReplacer;
+import io.prestosql.sql.planner.plan.JoinNode;
+import io.prestosql.sql.planner.plan.LimitNode;
+import io.prestosql.sql.planner.plan.OutputNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.SimplePlanRewriter;
+import io.prestosql.sql.planner.plan.SortNode;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static io.prestosql.sql.planner.plan.SimplePlanRewriter.rewriteWith;
+
+public class RemoveRedundantOrderByOptimizer
+        implements PlanOptimizer
+{
+    private static final Logger logger = Logger.get(RemoveRedundantOrderByOptimizer.class);
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        OrderByContext orderByContext = new OrderByContext(true);
+        return rewriteWith(new Rewriter(), plan, orderByContext);
+    }
+
+    private class Rewriter
+            extends SimplePlanRewriter<OrderByContext>
+    {
+        @Override
+        public PlanNode visitOutput(OutputNode node, RewriteContext<OrderByContext> context)
+        {
+            //Setting orderingUsedAbove to true
+            return context.defaultRewrite(node, new OrderByContext(true));
+        }
+
+        protected PlanNode visitPlan(PlanNode node, RewriteContext<OrderByContext> context)
+        {
+            boolean orderingUsedAbove = context.get().isOrderingUsedAbove();
+            if (node instanceof JoinNode || node instanceof AggregationNode || node instanceof LimitNode) {
```

Comment:
When `LIMIT` is order sensitive, it should get transformed into `TopN` (by `MergeLimitWithSort`). Of course, if we are to rely on this, we need to enforce some optimization order.

Reply:
The fundamental problem is that there's currently no way for the engine to reason about whether a sort is required (i.e., was it added by the planner to enforce some physical organization, was it a result of the user typing ORDER BY in their query in a place where it matters, or was it the result of the user typing ORDER BY in a place where it *doesn't* matter).

The SQL spec says that ORDER BY is relevant only for the immediate query expression that contains it. Since ORDER BY logically evaluates before FETCH FIRST (LIMIT) and OFFSET, those two operations in a query expression are sensitive the ORDER BY. If the ORDER BY was in a subquery, then it wouldn't. So:

```sql
SELECT * FROM t ORDER BY c LIMIT 1
```

and 

```sql
SELECT * FROM (
    SELECT * FROM t ORDER BY c
)
LIMIT 1
```

have the same query plan:

```
- Limit
   - Sort
     - TableScan
```

yet, the Sort is required in the first one but not in the second one.

Reply:
@martint what about
```sql
SELECT *, row_number() over () FROM (SELECT * FROM t ORDER BY c)
```

is the result of the query unspecified? or is the `row_number()` guaranteed to respect subquery ordering?

Note: even if nothing is _guaranteed_, people have used (and surely still use) this type of queries to emulate OFFSET, which we have only since recently. Removing Sort may break queries that were written the best way that was possible.

Reply:
@findepi, no because the ORDER BY applies only to the query expression that contains it (i.e., `SELECT * FROM t ORDER BY c`). For that, you need to do:

```sql
SELECT *, row_number() OVER (ORDER BY c) 
FROM (
    SELECT * FROM t)
)
```

Reply:
BTW, maybe we should start by having the analyzer emit a warning when there's an ORDER BY clause that's redundant. The two cases where people need to sort are supported by Presto in other ways. For window functions, via the ORDER BY clause within the window specification, as described above. For aggregation functions via the ORDER BY clause in the aggregation function:

```sql
SELECT array_agg(x ORDER BY y) FROM t
```

Reply:
Also, maybe a better approach for now is to have the planner skip adding sort nodes in those cases altogether. The analyzer could tag the ORDER BY as irrelevant and the planner would just ignore it.

Reply:
Apart from it we can also apply the same for sources of `Union`, `Intersection` and `Except` nodes ?

Reply:
Here's a prototype of the approach I described above: https://github.com/prestosql/presto/pull/818